### PR TITLE
RouterOS sync plugin

### PR DESCRIPTION
**Pull request by janprochazkacz**

### CODE REVIEW

File: plugins.json [hunk 1]
```diff
@@ -128,6 +128,23 @@
...
             "zipUrl": "https:\/\/github.com\/Ubiquiti-App\/UCRM-plugins\/raw\/master\/plugins\/revenue-report\/revenue-report.zip"
         },
+        {
+            "name": "routeros",
+            "displayName": "Mikrotik RouterOS",
+            "description": "Plugin for handling RouterOS devices.",
+            "url": "https:\/\/github.com\/Ubiquiti-App\/UCRM-plugins\/tree\/master\/plugins\/routeros",
+            "version": "1.0.0",
+            "ucrmVersionCompliancy": {
+                "min": null,
```
Comment (UBNT-Petr) [inline plugins.json]: `3.0.0` ?

Comment (keksa) [reply plugins.json]: nope, it actually can be just deleted, only the `unmsVersionCompliancy` should be present

File: plugins.json [hunk 2]
```diff
@@ -128,6 +128,23 @@
             "author": "Ubiquiti Networks, Inc.",
             "zipUrl": "https:\/\/github.com\/Ubiquiti-App\/UCRM-plugins\/raw\/master\/plugins\/revenue-report\/revenue-report.zip"
         },
+        {
+            "name": "routeros",
+            "displayName": "Mikrotik RouterOS",
```
Comment (keksa) [inline plugins.json]:
```suggestion
            "displayName": "Mikrotik RouterOS - suspension",
```

File: plugins.json [hunk 3]
```diff
@@ -128,6 +128,23 @@
...
+            "description": "Plugin for handling RouterOS devices.",
```
Comment (keksa) [inline plugins.json]:
```suggestion
            "description": "Plugin for handling suspension on RouterOS devices.",
```

File: plugins/routeros/README.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# Suspension plugin for Mikrotik RouterOS
+This plugin synchronizes list of suspended IP addresses from UCRM to RouterOS based router. 
+
+
+## Configuration
+To enable redirection to suspension page must be manually set _Filter Rules_ and _NAT_ firewall rules.
```
Comment (keksa) [inline plugins/routeros/README.md]:
```suggestion
To enable suspension page redirect, you must manually configure _Filter Rules_ and _NAT_ firewall rules.
```

Comment (keksa) [reply plugins/routeros/README.md]: maybe include some manual on how to do this?

File: plugins/routeros/src/manifest.json [hunk 5]
```diff
@@ -0,0 +1,47 @@
+{
+  "version": "1",
+  "information": {
+    "name": "routeros",
+    "displayName": "Mikrotik RouterOS",
```
Comment (keksa) [inline plugins/routeros/src/manifest.json]:
```suggestion
    "displayName": "Mikrotik RouterOS - suspension",
```

File: plugins/routeros/src/manifest.json [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    "description": "Plugin for handling RouterOS devices.",
```
Comment (keksa) [inline plugins/routeros/src/manifest.json]:
```suggestion
    "description": "Plugin for handling suspension on RouterOS devices.",
```

File: plugins/routeros/src/manifest.json [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    "url": "https://github.com/Ubiquiti-App/UCRM-plugins/tree/master/plugins/routeros",
+    "version": "1.0.0",
+    "ucrmVersionCompliancy": {
```
Comment (keksa) [inline plugins/routeros/src/manifest.json]: rm

File: plugins/routeros/src/manifest.json [hunk 8]
```diff
@@ -0,0 +1,47 @@
+{
+  "version": "1",
+  "information": {
+    "name": "routeros",
```
Comment (keksa) [inline plugins/routeros/src/manifest.json]: rename the plugin to `routeros-suspension` or something like that

File: plugins/routeros/src/src/Service/Suspender.php [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+        $this->setIpFirewallAddressList($this->findIpsFromNetwork($clientSiteIds));
+    }
+
+    private function findServicesToSuspend(): array
+    {
+        return $this->ucrmApi->get(
+            'clients/services',
+            [
+                'statuses' => [3],
```
Comment (keksa) [inline plugins/routeros/src/src/Service/Suspender.php]: const?

File: plugins/routeros/src/src/Service/Suspender.php [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            ]
+        );
+    }
+
+    private function findIpsFromNetwork(array $clientSiteIds): Set
+    {
+        $ipAddressses = new Set();
```
Comment (keksa) [inline plugins/routeros/src/src/Service/Suspender.php]: I think we talked about not using ext-ds for this plugin?

File: plugins/routeros/src/src/Service/UnmsApi.php [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        if (! $ucrmUrl) {
+            throw new ConfigurationException('UCRM URL is missing in plugin configuration.');
+        }
+
+        $config = (new PluginConfigManager())->loadConfig();
+        if (! ($config['unmsApiToken'] ?? false)) {
+            throw new ConfigurationException('UNMS API token is missing in plugin configuration.');
+        }
+
+        $ucrmApiUrl = sprintf(
```
Comment (keksa) [inline plugins/routeros/src/src/Service/UnmsApi.php]: `$unmsApiUrl` ?

File: plugins/routeros/src/src/Service/UnmsApi.php [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+            '%s/api/v2.1/',
+            preg_replace('~(crm\/?)\z~', 'nms', $ucrmUrl)
```
Comment (keksa) [inline plugins/routeros/src/src/Service/UnmsApi.php]: I think simple str_replace is more than enough here

File: plugins.json (hunk 1 above)
Comment (UBNT-Petr) [reply plugins.json]: @janprochazkacz ok, so then delete it. "min null and max null" looks like this is compatible with all UCRM versions

File: plugins/routeros-suspension/README.md [hunk 13]
```diff
@@ -0,0 +1,6 @@
+# Suspension plugin for Mikrotik RouterOS
+This plugin synchronizes list of suspended IP addresses from UCRM to RouterOS based router. 
+
+
+## Configuration
+To enable suspension page redirect, you must manually configure _Filter Rules_ and _NAT_ firewall rules.
```
Comment (keksa) [inline plugins/routeros-suspension/README.md]: previous CR is not resolved
> maybe include some manual on how to do this?

File: plugins/routeros-suspension/src/manifest.json [hunk 14]
```diff
@@ -0,0 +1,47 @@
+{
+  "version": "1",
+  "information": {
+    "name": "routeros-suspension",
+    "displayName": "Mikrotik RouterOS - suspension",
+    "description": "Plugin for handling suspension on RouterOS devices.",
+    "url": "https://github.com/Ubiquiti-App/UCRM-plugins/tree/master/plugins/routeros-suspension",
+    "version": "1.0.0",
+    "ucrmVersionCompliancy": {
```
Comment (keksa) [inline plugins/routeros-suspension/src/manifest.json]: previous CR not resolved, completely remove ucrmVersionCompliancy, it's only compatible with UNMS

File: plugins/routeros-suspension/src/src/Service/Suspender.php [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+                    'siteId' => $clientSiteId,
+                ]
+            );
+
+            foreach ($clientSiteIps as $clientSiteIp) {
+                $ipAddresses[] = $clientSiteIp;
+            }
+        }
+
+        return $ipAddresses;
```
Comment (keksa) [inline plugins/routeros-suspension/src/src/Service/Suspender.php]:
```suggestion
        return array_unique($ipAddresses);
```

File: plugins/routeros-suspension/src/src/Service/UnmsApi.php [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+        if (! $ucrmUrl) {
+            throw new ConfigurationException('UCRM URL is missing in plugin configuration.');
+        }
+
+        $config = (new PluginConfigManager())->loadConfig();
+        if (! ($config['unmsApiToken'] ?? false)) {
+            throw new ConfigurationException('UNMS API token is missing in plugin configuration.');
+        }
+
+        $unmsApiUrl = sprintf('%s/api/v2.1/', str_replace('crm/', 'nms', $ucrmUrl));
```
Comment (keksa) [inline plugins/routeros-suspension/src/src/Service/UnmsApi.php]: This looks like a bug, even though it probably isn't. Can we maybe do something like this?
```suggestion
        $unmsApiUrl = sprintf('%s/api/v2.1/', rtrim(str_replace('crm/', 'nms/', $ucrmUrl), '/'));
```

File: plugins/routeros-suspension/src/src/Service/Suspender.php [hunk 17]
```diff
@@ -0,0 +1,399 @@
...
+        $natSrcJumpRules = [
+            'first_srcnat',
+            'general_srcnat',
+            'range_srcnat',
+            'last_srcnat',
+        ];
+
+        $rules = [];
+
+        //add jump rules
```
Comment (Piskvor) [inline plugins/routeros-suspension/src/src/Service/Suspender.php]:
```suggestion
        // add jump rules
```

**Pull request merged**